src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferenceCache.cs

```diff
@@ -17,24 +17,44 @@
 
 namespace Microsoft.CodeAnalysis.FindSymbols;
 
+/// <summary>
+/// Caches information find-references needs associated with each document.  Computed and cached so that multiple calls
+/// to find-references in a row can share the same data.
+/// </summary>
 internal sealed class FindReferenceCache
 {
-    private static readonly ConditionalWeakTable<SemanticModel, FindReferenceCache> s_cache = new();
+    private static readonly ConditionalWeakTable<Document, AsyncLazy<FindReferenceCache>> s_cache = new();
 
-    public static FindReferenceCache GetCache(SemanticModel model)
-        => s_cache.GetValue(model, static model => new(model));
+    public static async ValueTask<FindReferenceCache> GetCacheAsync(Document document, CancellationToken cancellationToken)
+    {
+        var lazy = s_cache.GetValue(document, static document => AsyncLazy.Create(ComputeCacheAsync, document));
+        return await lazy.GetValueAsync(cancellationToken).ConfigureAwait(false);
 
-    private readonly SemanticModel _semanticModel;
+        static async Task<FindReferenceCache> ComputeCacheAsync(Document document, CancellationToken cancellationToken)
+        {
+            // Find-Refs is not impacted by nullable types at all.  So get a nullable-disabled semantic model to avoid
+            // unnecessary costs while binding.
+            var model = await document.GetRequiredNullableDisabledSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            return new(document, model, root);
+        }
+    }
+
+    public readonly Document Document;
+    public readonly SemanticModel SemanticModel;
+    public readonly SyntaxNode Root;
```

Comment:
had this hold onto Doc/SemanticModel/Root to ease up on hte number of arguments i'm passing around later.